fvm/state/accounts.go

```diff
@@ -324,19 +322,120 @@ func (a *Accounts) setContractNames(contractNames contractNames, address flow.Ad
 		return nil
 	}
 
-	a.ledger.Set(string(address.Bytes()), string(address.Bytes()), keyContractNames, newContractNames)
+	return a.setValue(address, true, keyContractNames, newContractNames)
+}
+
+// GetStorageUsed returns the amount of storage used in bytes by this account
+func (a *Accounts) GetStorageUsed(address flow.Address) (uint64, error) {
+	storageUsedRegister, err := a.getValue(address, false, keyStorageUsed)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(storageUsedRegister) != uint64StorageSize {
+		return 0, fmt.Errorf("account %s storage used is not initialized or not initialized correctly", address.Hex())
+	}
+
+	storageUsed, _, err := utils.ReadUint64(storageUsedRegister)
+	if err != nil {
+		return 0, err
+	}
+	return storageUsed, nil
+}
+
+func (a *Accounts) setStorageUsed(address flow.Address, used uint64) error {
+	usedBinary := utils.Uint64ToBinary(used)
+	return a.setValue(address, false, keyStorageUsed, usedBinary)
+}
 
+// GetValue returns a value stored in address' storage
+func (a *Accounts) GetValue(address flow.Address, key string) (flow.RegisterValue, error) {
+	return a.getValue(address, false, key)
+}
+
+func (a *Accounts) getValue(address flow.Address, isController bool, key string) (flow.RegisterValue, error) {
+	if isController {
+		return a.ledger.Get(string(address.Bytes()), string(address.Bytes()), key)
+	}
+	return a.ledger.Get(string(address.Bytes()), "", key)
+}
+
+// SetValue sets a value in address' storage
+func (a *Accounts) SetValue(address flow.Address, key string, value flow.RegisterValue) error {
+	return a.setValue(address, false, key, value)
+}
+
+func (a *Accounts) setValue(address flow.Address, isController bool, key string, value flow.RegisterValue) error {
+	err := a.updateRegisterSizeChange(address, isController, key, value)
+	if err != nil {
+		return fmt.Errorf("failed to update storage used by key %s on account %s: %w", key, address, err)
+	}
+
+	if isController {
+		a.ledger.Set(string(address.Bytes()), string(address.Bytes()), key, value)
+	} else {
+		a.ledger.Set(string(address.Bytes()), "", key, value)
+	}
 	return nil
 }
 
+func (a *Accounts) updateRegisterSizeChange(address flow.Address, isController bool, key string, value flow.RegisterValue) error {
+	if key == keyStorageUsed {
+		// size of this register is always uint64StorageSize
+		// don't double check this to save time and prevent recursion
+		return nil
+	}
+	oldValue, err := a.getValue(address, isController, key)
+	if err != nil {
+		return err
+	}
+
+	sizeChange := int64(len(value) - len(oldValue))
```

Comment:
This is not using the key part as part of storage used yet. I want to add that in a different PR as it needs a few more changes.